overlord/snapstate/backend/snapdata_test.go

```diff
@@ -117,3 +117,33 @@ func (s *snapdataSuite) TestRemoveSnapDataDir(c *C) {
 	// the snap-name directory is gone now too
 	c.Assert(osutil.FileExists(varBaseData), Equals, false)
 }
+
+func (s *snapdataSuite) TestRemoveUserSnapDir(c *C) {
+	homedir := filepath.Join(s.tempdir, "home", "user")
+
+	for _, opts := range []*dirs.SnapDirOptions{
+		{HiddenSnapDataDir: true},
+		{HiddenSnapDataDir: false},
+		nil,
+	} {
+		snapHomeDir := dirs.UserHomeSnapDir
+		if opts != nil && opts.HiddenSnapDataDir {
+			snapHomeDir = dirs.HiddenSnapDataHomeDir
+		}
+
+		mkdir := func(dir string) { c.Assert(os.MkdirAll(dir, 0700), IsNil) }
+		mkdir(filepath.Join(homedir, snapHomeDir, "hello", "common"))
+		mkdir(filepath.Join(homedir, snapHomeDir, "hello", "x1"))
+		mkdir(filepath.Join(dirs.GlobalRootDir, "root", snapHomeDir, "hello", "common"))
+		mkdir(filepath.Join(dirs.GlobalRootDir, "root", snapHomeDir, "hello", "x1"))
+
+		c.Assert(s.be.RemoveUserSnapDir("hello", opts), IsNil)
+
+		for _, d := range []string{
+			filepath.Join(homedir, snapHomeDir, "hello"),
+			filepath.Join(dirs.GlobalRootDir, "root", snapHomeDir, "hello"),
+		} {
+			c.Assert(osutil.FileExists(d), Equals, false)
```

Comment:
It would be nice to check that the parent directory still exists, just to make sure that the implementation doesn't accidentally remove the whole filesystem :-)